nexus/types/src/external_api/params.rs

```diff
@@ -18,6 +18,88 @@ use serde::{
 use std::{net::IpAddr, str::FromStr};
 use uuid::Uuid;
 
+#[derive(Deserialize, JsonSchema)]
+#[serde(untagged)]
+pub enum ProjectSelector {
+    ProjectId {
+        project_id: Uuid,
+    },
+    ProjectAndOrgId {
+        project_name: Name,
+        organization_id: Uuid,
+    },
+    ProjectAndOrg {
+        // FIXME: There's a bug in schemars or serde which causes project_name to be emitted twice
+        #[schemars(skip)]
+        project_name: Name,
+        organization_name: Name,
+    },
+    None {},
```

Comment:
Does this `None` need to be built into the enum because of how deserialization works? Otherwise it seems like a natural spot for `Option` or `Result`. Alternatively, I feel like in the `None` case the endpoint should automatically 400 at the parsing, which would mean you don't need the `None` in the enum itself.

Reply:
So an untagged enum tries to match all of the options and errors out if it can't find a match. I _think_ wrapping it in an option might be viable. I didn't take that tack because we always need to match against the enum anyway and having a None option (that essentially means there were not matches in the untagged enum) makes the handling easier. 

If a route has a path parameter and that path parameter is not an ID then the selector having a matching arm is required. Given that two types are involved here the optional isn't really valid. In this case we can actually document `None` as a part of the API where it's understood at what point you shouldn't include a selector.